### ISSUE TITLE
Support functions with zero arguments

### ISSUE THREAD

**Issue by dylanede** [dtolnay/reflect#13]
At the moment, the `library` macro will not compile if a function is declared that takes zero arguments. This is because `[T; 0]` does not implement `Array`. I see that the implementation of `Function::invoke` relies on there being a first argument to the function, so I'm not sure how to easily modify this to allow functions with zero arguments.

**Comment by 8BitMate** [comment]
I've looked into this as well. The reason ```Function::invoke``` won't work for zero arguments, is that the ```invoke``` function needs a ```WipFunction``` in order to know which function it is being invoked from. ```Value```elements contains a ```WipFunction``` reference, but if no values are supplied, then there is no way of getting the correct ```WipFunction``` for ```Function::invoke```. From which function another function is invoked is not something the ```library``` macro can know in advance, so it has to get this information from somewhere.

I don't know of a very elegant way of solving this, but one possible solution is to make a ```Function::invoke_empty``` with this signature:
```Rust
pub fn invoke_empty<'a>(&self, f: MakeFunction<'a>) -> Value<'a>
```
and then let the ```library``` macro generate ```INVOKE``` functions along the line of this:
```Rust
pub fn INVOKE<'a>(self, f: _reflect::MakeFunction<'a>) -> _reflect::Value<'a> {
    _reflect::runtime::RuntimeFunction::SELF(self).invoke_empty(f)
}
```
for functions with no parameters.

If you wanted to invoke ```String::new``` you could do it like this:
```Rust
RUNTIME::std::string::String::new.INVOKE(f.clone())
```
where f is a ```MakeFunction```. If we derive ```Copy``` for ```MakeFunction``` it becomes even simpler
```Rust
RUNTIME::std::string::String::new.INVOKE(f)
```
One drawback to this is that f is not an actual argument to the function that is being invoked, but only the context in which it is called, so it can be confusing for users of the library, but I haven't come up with a better solution.

**Comment by dtolnay** [comment]
Maybe we could keep track of the current MakeFunction in a [thread_local](https://doc.rust-lang.org/std/thread/struct.LocalKey.html). I don't see us needing to support multiple threads inside the code generator, so there will only ever be one thread and it can get the MakeFunction that way. Threads in a proc macro would be extremely unusual anyway because none of the proc_macro API types implement Send.

```rust
// make_function would stick the new MakeFunction in thread_local
// within the reflect crate, anything inside of debug_fmt would read
// it from there.
block.make_function(RUNTIME::std::fmt::Debug::fmt, debug_fmt);
```

**Comment by 8BitMate** [comment]
Fixed in a515c1d11d1bd9699a0e705dad878cb44b331190 and a024cbeb16cc01f70c18da5784595551dce6e67a
